data/blog/darwin-spelunking.mdx

```diff
@@ -0,0 +1,46 @@
+  - 'oss'
+authors: ['mihai']
+images: ['/images/darwin-spelunking/social.png']
+summary: 'Spelunking through the Darwin kernel and userland helps Tailscale work better on macOS and iOS.'
+---
+
+# Improving Tailscale via Apple’s Open Source
+
+[Daniel Jalkut’s](https://redsweater.com/about/DanielJalkut.html) recent [Spelunking Apple’s Open Source](https://bitsplitting.org/2023/03/17/spelunking-apples-open-source/) blog post struck a very familiar chord with me. Some of my most satisfying times at my previous employer were spent [poring through WebKit internals](https://blog.persistent.info/search/label/WebKit). At Tailscale the pattern has continued, this time more focused on [Darwin](<https://en.wikipedia.org/wiki/Darwin_(operating_system)>) (the operating system at the core of macOS and iOS), its [kernel](https://en.wikipedia.org/wiki/XNU) and [userland](https://en.wikipedia.org/wiki/User_space_and_kernel_space) tools. As an example of the power of being able to peek under the hood of the operating system, here’s a recent debugging story from the Tailscale 1.36 release:
+
+We were investigating two network-interface related bugs – one where Tailscale traffic was [looping on itself](https://github.com/tailscale/tailscale/issues/5156) (instead of being passed through to the actual physical network connection) and another where Tailscale was [sending traffic over the phone’s cellular interface](https://github.com/tailscale/tailscale/issues/6565) (even though Wi-Fi was available). The [fix](https://github.com/tailscale/tailscale/pull/6566) for both of these involved more correctly “binding” Tailscale’s network requests to the active network interface (and re-binding them when it changes).
+
+Normally figuring out what the active network interface is pretty straightforward – it is the interface that is [associated with the default route](https://github.com/tailscale/tailscale/blob/9c2ad7086c0daef9c914ef1e59d0142b50f789f3/net/interfaces/interfaces_bsd.go#L38-L51). However, when Tailscale is configured to [use an exit node](https://tailscale.com/kb/1103/exit-nodes/), it becomes the default route, thus putting us back in danger of looping traffic. We want to know what the interface would be if Tailscale was not present (but without actually disabling it, since that is [disruptive](https://bugs.chromium.org/p/chromium/issues/detail?id=1076619)). I initially had an approach that [lived in our closed-source repo](https://github.com/tailscale/tailscale/blob/42c3052006141aab7d2573b2019c996641384014/net/interfaces/interfaces_bsd.go#L85-L93) and relied on private APIs (that I should probably [plead the 5th](https://en.wikipedia.org/wiki/Fifth_Amendment_to_the_United_States_Constitution#Self-incrimination) instead of discussing further). Worse yet, we started to get reports that users were [having](https://github.com/tailscale/tailscale/issues/7184) [issues](https://github.com/tailscale/tailscale/issues/7188), we were sometimes detecting the wrong interface.
```

Comment:
Nit but we should use an — em dash here. Our observed convention is with a space around them, as you have here.